core/touch.js

```diff
@@ -0,0 +1,263 @@
+      Blockly.longStop_();
+      workspace.dragMode_ = Blockly.DRAG_FREE;
+    }
+    e.stopPropagation();
+    e.preventDefault();
+  }
+};
+
+/**
+ * Clear the touch identifier that tracks which touch stream to pay attention
+ * to.  This ends the current drag/gesture and allows other pointers to be
+ * captured.
+ */
+Blockly.Touch.clearTouchIdentifier = function() {
+  Blockly.Touch.touchIdentifier_ = null;
+};
+
+/**
+ * Decide whether Blockly should handle or ignore this event.
+ * Mouse and touch events require special checks because we only want to deal
+ * with one touch stream at a time.  All other events should always be handled.
+ * @param {!Event} e The event to check.
+ * @return {boolean} True if this event should be passed through to the
+ *    registered handler; false if it should be blocked.
+ */
+Blockly.Touch.shouldHandleEvent = function(e) {
+  return !Blockly.Touch.isMouseOrTouchEvent(e) ||
+      Blockly.Touch.checkTouchIdentifier(e);
+};
+
+/**
+ * Check whether the touch identifier on the event matches the current saved
+ * identifier.  If there is no identifier, that means it's a mouse event and
+ * we'll use the identifier "mouse".  This means we won't deal well with
+ * multiple mice being used at the same time.  That seems okay.
+ * If the current identifier was unset, save the identifier from the
+ * event.  This starts a drag/gesture, during which touch events with other
+ * identifiers will be silently ignored.
+ * @param {!Event} e Mouse event or touch event.
+ * @return {boolean} Whether the identifier on the event matches the current
+ *     saved identifier.
+ */
+Blockly.Touch.checkTouchIdentifier = function(e) {
+  var identifier = (e.changedTouches && e.changedTouches[0] &&
+        e.changedTouches[0].identifier != undefined &&
+        e.changedTouches[0].identifier != null) ?
+        e.changedTouches[0].identifier : 'mouse';
+
+  // if (Blockly.touchIdentifier_ )is insufficient because android touch
+  // identifiers may be zero.
+  if (Blockly.Touch.touchIdentifier_ != undefined &&
+      Blockly.Touch.touchIdentifier_ != null) {
+    // We're already tracking some touch/mouse event.  Is this from the same
+    // source?
+    return Blockly.Touch.touchIdentifier_ == identifier;
+  }
+  if (e.type == 'mousedown' || e.type == 'touchstart') {
+    // No identifier set yet, and this is the start of a drag.  Set it and
+    // return.
+    Blockly.Touch.touchIdentifier_ = identifier;
+    return true;
+  }
+  // There was no identifier yet, but this wasn't a start event so we're going
+  // to ignore it.  This probably means that another drag finished while this
+  // pointer was down.
+  return false;
+};
+
+/**
+ * Set an event's clientX and clientY from its first changed touch.  Use this to
+ * make a touch event work in a mouse event handler.
+ * @param {!Event} e A touch event.
+ */
+Blockly.Touch.setClientFromTouch = function(e) {
+  if (e.type.indexOf('touch') == 0) {
```

Comment:
If e.type doesn't start with 'touch', this is quite inefficient since JS needs to work its way all the way through the string using [KMP](https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm).  A better approach is to use:
e.type.lastIndexOf('touch', 0) == 0
This exits instantly if the match isn't found since it runs off the left of the string.

Or better yet, just use goog.string.startsWith.